splitgraph/madatdata
# Upgrade TypeScript (and related) to 5.0, and start using `package.json` conditional exports

Somewhat surprisingly, this all seems to be working.

Basically:

- Upgrade TypeScript to latest
- Opt out of `verbatimModuleSyntax` (but continue using `isolatedModules: true`)
- Upgrade tsc-multi to latest https://github.com/tommy351/tsc-multi/issues/19
- Continue targeting CJS outputs! :) 
- Use `exports` field of `package.json`, and specifically "conditional exports," so that published modules point to CJS or ESM (only using `es2020` for ESM atm, although we build `esnext`) based on consumer, e.g.:

```js
// print out the resolved path of a module imported via the module system
// (this flag is just necessary to get import.meta.resolve to print out the file)
// node --experimental-import-meta-resolve blah.mjs
console.log(await import.meta.resolve("@madatdata/react"));

// outputs:
// file:///path/to/madatdata/examples/react-nextjs-basic-hooks/node_modules/@madatdata/react/build/es2020/index.mjs
```

and

```js
// node blah.js
console.log(require.resolve("@madatdata/react"));

// outputs:
// /path/to/madatdata/examples/react-nextjs-basic-hooks/node_modules/@madatdata/react/build/es2020-commonjs/index.cjs
```

- Use a custom conditional export called `dev` which is not consumed by external tools (and therefore will not interfere with downstream packages), and use the new TypeScript 5.0 feature `customConditions` to make sure that TypeScript can resolve using `moduleResolution: nodenext` by seeing `index.ts` (instead of something in `./build` for each package). Similarly, use a config flag `resolutions.conditions` in vitest, which passes it to vite, which passes it to `node --conditions` (see: https://github.com/vitest-dev/vitest/actions/runs/3831432535)
- Upgrade vitest to latest
- Upgrade msw to latest and drop our hacky patch
- Build latest versions of all packages and consume them with examples to make sure they still work

Once this is merged I'll publish the latest versions to npm. 

Unfortunately I think skypack will still be broken, since it doesn't install dependencies of a package. But we'll see, maybe it will magically work too (doubt it).
